staging/kos/pkg/controllers/subnet/validator.go

```diff
@@ -0,0 +1,458 @@
+	defer v.queue.Done(subnet)
+	requeues := v.queue.NumRequeues(subnet)
+	if err := v.processSubnet(subnet); err != nil {
+		glog.Warningf("Failed processing %s, requeuing (%d earlier requeues): %s.", subnet, requeues, err.Error())
+		v.queue.AddRateLimited(subnet)
+		return
+	}
+	glog.V(4).Infof("Finished %s with %d requeues.", subnet, requeues)
+	v.queue.Forget(subnet)
+}
+
+func (v *Validator) processSubnet(subnetNSN k8stypes.NamespacedName) error {
+	subnet, err := v.subnetLister.Subnets(subnetNSN.Namespace).Get(subnetNSN.Name)
+
+	if err != nil && !k8serrors.IsNotFound(err) {
+		glog.Errorf("subnet lister failed to lookup %s: %s", subnetNSN, err.Error())
+		// This should never happen. No point in retrying.
+		return nil
+	}
+
+	if k8serrors.IsNotFound(err) {
+		v.processDeletedSubnet(subnetNSN)
+		return nil
+	}
+
+	return v.processExistingSubnet(subnet)
+}
+
+func (v *Validator) processDeletedSubnet(s k8stypes.NamespacedName) {
+	v.clearStaleRV(s)
+
+	rivals := v.clearConflictsCache(s)
+
+	// Enqueue old rivals so that they can be re-validated: they might no longer
+	// have conflicts as this subnet has been deleted.
+	for _, r := range rivals {
+		v.queue.Add(r)
+	}
+}
+
+func (v *Validator) processExistingSubnet(s *netv1a1.Subnet) error {
+	ss, parsingErrs := subnet.NewSummary(s)
+	if len(parsingErrs) > 0 {
+		return parsingErrs
+	}
+
+	if v.subnetIsStale(ss.NamespacedName, s.ResourceVersion) {
+		return nil
+	}
+
+	// If we're here s might have been created or updated in a way that affects
+	// validation. We need to update its conflicts cache accordingly and
+	// reconsider old rivals because they might no longer be in conflict with
+	// s.
+	oldRivals := v.updateConflictsCache(ss)
+	for _, r := range oldRivals {
+		v.queue.Add(r)
+	}
+
+	// When a subnet is validated, in case of conflicts it is added to the
+	// conflicts caches of its rival subnets, so that when such rivals are
+	// deleted it can be revalidated. This is useless for a subnet that has
+	// already passed validation. Also, if this controllers sees
+	// s.Status.Validated to be true, there's no way that such value can refer
+	// to a previous, stale version of the subnet, because the API server clears
+	// s.Status.Validated when a subnet spec is updated in a way that requires
+	// revalidation. Hence, we can simply stop processing here if
+	// s.Status.Validated is true.
+	// ? Maybe we should make it clearer somewhere (e.g. API types comments and
+	// doc) that the API server clears s.Status.Validated when a subnet spec is
+	// updated in a way that requires revalidation.
+	if s.Status.Validated {
+		return nil
+	}
+
+	// Retrieve all the other subnets. Doing a live list as opposed to a
+	// cache-based one (through the informer) prevents race conditions that can
+	// arise in case of multiple validators running.
+	allSubnets, err := v.netIfc.Subnets(k8scorev1api.NamespaceAll).List(k8smetav1.ListOptions{})
```

Comment:
We can and should make this much more efficient by restricting the list to those subnets with the same VNI.  This will require adding that field selector to those supported.

Reply:
Fixed in 5254ee4.